rpc/src/responses/block_ids_in_chain.rs

```diff
@@ -19,6 +20,66 @@ impl Response for BlocksInChainResponse {
 #[cfg(test)]
 mod test {
     use super::*;
+    use serde::{Deserialize, Serialize};
+
+    /// Mock struct that represents the
+    /// object expected to be found in
+    /// a response.
+    #[derive(Serialize, Deserialize)]
+    struct MockBlockInChain {
+        block_id: String,
+    }
+
+    struct BlocksInChainMockResponse {
+        pub mock_block_ids: json_array::JsonArray<MockBlockInChain>,
+    }
+
+    impl Response for BlocksInChainMockResponse {
+        /// Parses a response string in the form
+        /// `"[[{"block_id":"alpha_numeric_block_id_string"}], ["..."]]"` or
+        /// `"[{"block_id":"alpha_numeric_block_id_string"}, ...]"` into a
+        /// [`BlocksInChainMockResponse`](Self).
+        fn from_response_str(response: &str) -> Result<Self, ParseError> {
+            let mock_block_ids = json_array::JsonArray::from_str(response)?;
+            Ok(Self { mock_block_ids })
+        }
+    }
+
+    #[test]
+    fn get_blocks_in_chain_from_object_response_expecting_string_fails() {
+        let mock_response = r#"[[{"block_id": "1"}]]"#;
+
+        let blocks_response = BlocksInChainResponse::from_response_str(mock_response);
+        assert!(blocks_response.is_err());
+    }
+
+    #[test]
+    fn get_blocks_in_chain_from_object_response_expecting_object_ok() {
+        let mock_response = r#"[[{"block_id": "1"}]]"#;
+
+        let blocks_response = BlocksInChainMockResponse::from_response_str(mock_response);
+        assert!(blocks_response.is_ok());
+
+        let blocks = blocks_response.unwrap().mock_block_ids.into_vec();
+        assert!(blocks.len() == 1);
+
+        let single_block = &blocks[0];
+        assert!(single_block.block_id == "1");
+    }
+
+    #[test]
+    fn get_blocks_in_chain_from_flat_array_object_response_expecting_object_ok() {
+        let mock_response = r#"[{"block_id": "2"}]"#;
```

Comment:
If the parsing format is supposed to be `[[{"block_id": "num"}], [{...}]]`, should we even _allow_ for a parsing of an erroneous format? AFAIK even if there is only a single entry in the array, the RPC spec will return it in a nested array (because of the row/column format).

It might do more harm than good to accept and parse RPC response strings that are even _slightly_ off spec, as they might signify a greater problem that the end user might want to explicitly handle.

This is all (obviously) only if the RPC always returns a `[[]]` instead of ever a `[]`, even with only one entry.

Reply:
You are totally right, I tried testing JsonArray in a file that is specific to one RPC endpoint so this doesn't make sense at all. We can save tests like this one for another file whose RPC return matches the `[[{"block_id": "1"}]]` form.